contracts/common/EtherRouterCreate3.sol

```diff
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+/*
+  This file is part of The Colony Network.
+
+  The Colony Network is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  The Colony Network is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with The Colony Network. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+pragma solidity 0.8.25;
+
+import { EtherRouter } from "./EtherRouter.sol";
+
+contract EtherRouterCreate3 is EtherRouter {
```

Comment:
Tried writing this out a couple of times, gave up, and ended up deciding it needed a diagram. 

![Colony Network](https://github.com/JoinColony/colonyNetwork/assets/311812/75e9980f-b5c3-4c68-a8aa-bcb6b0adc83f)

This all happens in a single transactions when creating a colony.

The first thing that CreateX does is create a temporary deployer contract. This is deployed via the [CREATE2](https://eips.ethereum.org/EIPS/eip-1014) opcode, which means the resulting address is a function of:

* The creator address (which is always the CreateX address `0xba5ed...ba5ed`)
* A salt, which we pass
* The initcode of the contract, which is fixed by CreateX

That temporary deployment contract then deploys the contract that will become the colony via the CREATE opcode. That means the resulting address is a function of the creator's address (i.e. the temporary deployment contract's address) and the account's nonce; as we are using a new (temporary) deployment contract, the nonce is 0. 

The temporary contract then self-destructs. The init call is then sent from CreateX. This is the crux of the answer to your question, as to why we can't use the normal EtherRouter. When the EtherRouter is deployed, it will have its owner set to the address of the temporary deployer contract, but there's no way to call a `setOwner` from that address. `EtherRouterCreate3` sets the owner straight away to CreateX, which can then make the `setOwner` call to set the owner to ColonyNetwork.

We _could_ hardcode the ColonyNetwork address here, and not make any init call at all. I went against that primarily because it would mean that we would need to have the same address for the ColonyNetwork locally, on testnets, and in production, which I am wary of. It also gives us the option to have something that is not the ColonyNetwork use this code for deterministic deployments in the future. If we end up being comfortable with the idea of hard-coding the ColonyNetwork address here, this is something we can change in the future (because, by using the `CREATE3` pattern, the address of the resulting colony is independent of the bytecode of EtherRouter).

Reply:
Interesting. Why create the temporary deployer vs. having CreateX deploy directly?

Reply:
I'm assuming the deployer is the proxy referenced [here?](https://github.com/pcaversaccio/createx/blob/main/src/CreateX.sol#L686)

Reply:
For posterity: The diagram is wrong in that the proxy contract does not selfdestruct. I'm not sure why, though.